src/thinblock.cpp

```diff
@@ -647,8 +647,7 @@ bool CXThinBlock::HandleMessage(CDataStream &vRecv, CNode *pfrom, string strComm
     }
 
     // Send expedited block without checking merkle root.
-    if (!IsRecentlyExpeditedAndStore(inv.hash))
-        SendExpeditedBlock(thinBlock, nHops, pfrom);
+    SendExpeditedBlock(thinBlock, nHops, pfrom);
```

Comment:
The SendExpeditedBlock(CXThinBlock &thinBlock,...) function does not call "IsRecentlyExpeditedAndStore".  Only the SendExpeditedBlock(const CBlock &block) function does so.  So this change might cause an expedited forwarding loop.  I know what you're probably going to say though, and I agree: the original code is flawed because same-named functions should have the same semantics.  So the right solution is not to revert your change, but to call IsRecentlyExpeditedAndStore in the thin block variant of SendExpeditedBlock.  However, this is not straight forward since the CBlock variant calls the thin block variant.  We need IsRecently... in the SendExpedited...(CBlock) because we don't want to go through all the trouble of producing a CXThinBlock just to decide not to use it.  But we cannot call IsRecently twice...  So it seems like we need to create a new static helper function that both SendExpeditedBlock functions call.  This helper function takes a CXThinBlock and sends without checking.   And both SendExpeditedBlock function call IsRecently... and then the helper function.

Reply:
Ok, thanks for the heads up. Yes, good that you caught this and I have to say that was a trap too easy to fall into :D

I'll see what I can do.